### ISSUE TITLE
eliminating repetition of parameters?

### ISSUE THREAD

**Issue by RichMorin** [drolbr/Overpass-API#453]
I'd like to reduce the size of my Overpass QL queries, by eliminating repetition of parameters.  Am I missing some sort of parameter handling syntax (e.g., a way to store arguments in variables)?  Here are a few examples, for purposes of discussion:

Get various types of items around a specified location:

    ( node(around:...)[amenity=*];
       way(around:...)[k=shop,v=*];
    ); out; # Output the result.

Get all nodes, ways, and relations within a polygon:

    ( node( poly:"..." ); way( poly:"..." ); relation( poly:"..." );
      ...
    ); out;

Get all nodes, ways, and relations within a bounding box:

  (     node("...", "...", "...", "...");
         way("...", "...", "...", "...");
    relation("...", "...", "...", "...");
  );
  (  ._;  # Recall the retrieved data as the "default set".
     <<;  # Recurse up to resolve unsatisfied ID references.
  ); out; # Output the result.

I'd rather not repeat the bounding boxes three times.  I realize that I can specify a bounding box for the query as a whole, but I worry about whether this might constrain the upward recursion (<<) operator.

**Comment by mmd-osm** [comment]
This seems like a duplicate of #150

**Comment by mmd-osm** [comment]
> Please note my support for having this feature in QL

Right, that's exactly what is stated in https://github.com/drolbr/Overpass-API/issues/150#issue-48197983 - I just added the comment about Overpass Turbo back then as it is relevant for most users.

> repeating the parameters seems inelegant and may even be slowing things down a bit

As you're creating queries programatically anyway, that's not much of an issue, then. Performance for sure isn't one. Also, the alternative mentioned in #150 is merely syntactic sugar and not really changing the query execution itself in any way.

BTW: Regarding your last example, there's the alternative of a  _global bounding box_. which would automatically apply to all statements.

**Comment by RichMorin** [comment]
Would the global bounding box be applied to the upward recursion (<<) operator?  If so, that would kind of defeat the purpose of the recursion.

**Comment by mmd-osm** [comment]
> Would the global bounding box be applied to the upward recursion (<<) operator

At the moment, recursion should disregard the global bounding box. 

> that would kind of defeat the purpose of the recursion.

You could argue either way, that's why #402 is still open and labeled as 'bug'.

**Comment by RichMorin** [comment]
#402 seems to indicate that recursion will continue to disregard the global bounding box. If this is the case, I'd be happy to use the GBB.  If not, not. What would you advise?

**Comment by mmd-osm** [comment]
Well, given that @drolbr stated that _it is intended that the recurse statement does not use the global bounding box_, it is rather unlikely to have this incompatible change as an outcome of #402.

BTW: I just added another example to #402 to illustrate how to filter a recursion result by global bbox.

**Comment by RichMorin** [comment]
So, I now have an answer for the case of bounding boxes.  That's a big win; I'm already using it. However, I still don't have a way to avoid repeating the arguments if I want to search for nodes, ways, and rels using around or poly.

I'd like a way to name an argument (eg, a poly string) and then use the name in the node, way, and ref operators. Alternatively, it might be useful to have a way to apply more than one operator to a single set of arguments.  Am I missing something?

**Comment by mmd-osm** [comment]
> I'd like a way to name an argument (eg, a poly string) and then use the name in the node, way, and ref operators.

What's in your poly? If that's an existing OSM way or relation, why not use that instead. Also, you could reduce the poly string size by reducing the number of decimal places, reducing points via douglas-peucker, ... After all, if you're sending your query via HTTP POST, I wouldn't spend too much time tweaking the last bits here. You can easily send an 1MB query that way.

> it might be useful to have a way to apply more than one operator to a single set of arguments

You can add as many filters to a query as you like, e.g. `way[highway](around:10)(poly:12.2,...);`

>  Am I missing something?

I don't know what your use case is, so general advise is a bit difficult. Do you have an example query you could to share via Overpass Turbo shortlink?

**Comment by mmd-osm** [comment]
> I want to search for nodes, ways, and rels using around or poly.

Forthcoming 0.7.55 release will provide an "nwr" statement in addition to node, way, rel and area. This way you can avoid repeating the same arguments by querying node, way and relations at the same time. Example:

```C
area[name="Bonn"];
node(area)[highway=bus_stop];
nwr(around:100)[amenity=cinema];
out center;
```

I think this issue can be closed, then.
